src/Servers/Kestrel/Core/src/Middleware/TlsListenerMiddleware.cs

```diff
@@ -0,0 +1,124 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Buffers;
+using System.Diagnostics;
+using Microsoft.AspNetCore.Connections;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Core.Middleware;
+
+internal sealed class TlsListenerMiddleware
+{
+    private readonly ConnectionDelegate _next;
+    private readonly Action<ConnectionContext, ReadOnlySequence<byte>> _tlsClientHelloBytesCallback;
+
+    public TlsListenerMiddleware(ConnectionDelegate next, Action<ConnectionContext, ReadOnlySequence<byte>> tlsClientHelloBytesCallback)
+    {
+        _next = next;
+        _tlsClientHelloBytesCallback = tlsClientHelloBytesCallback;
+    }
+
+    /// <summary>
+    /// Sniffs the TLS Client Hello message, and invokes a callback if found.
+    /// </summary>
+    internal async Task OnTlsClientHelloAsync(ConnectionContext connection)
+    {
+        var input = connection.Transport.Input;
+
+        while (true)
```

Comment:
This is an infinite loop if someone sends 7+ bytes but not the complete client hello and then closes the connection. You need an exit condition, see Infinite Loop in https://learn.microsoft.com/dotnet/standard/io/pipelines#pipereader-common-problems

Reply:
added a test for 9 bytes message, and added an exit condition for `NotEnoughData` if `ReadResult.IsCompleted` is true:
```
if (parseState == ClientHelloParseState.NotEnoughData)
{
    // if no data will be added, and we still lack enough bytes
    // we can't block in a loop, so just exit
    if (result.IsCompleted)
    {
        break;
    }

    continue;
}
```